File: voom-fe/src/app/shared/rest/ride/ride.api.ts
import { Api } from '../api';
import { ApiClient } from '../api-client';
import {
  DriverSummaryDto, OngoingRideDto, RatingRequestDto,
  RideCancellationDto, RideHistoryDto,
  RidePanicDto, RideReportRequestDto, RideRequestDto, RideRequestResponseDto, RideResponseDto,
  RideStopDto, StartRideDto
} from './ride.model';
import { RequestConfig } from '../rest.model';
import { ApiResponse } from '../rest.model';
import {ActiveRideDto, StartScheduledRideDto} from '../../../main-shell/user-pages/home/home.api';
import { Observable } from 'rxjs/internal/Observable';

export class RideApi extends Api {

  constructor(apiClient: ApiClient) {
    super(apiClient);
  }

  cancelRide(rideId: number, body: RideCancellationDto) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.post<RideCancellationDto, RideResponseDto>(
      `/api/rides/${rideId}/cancel`,
      body,
      config
    );
  }

  stopRide(rideId: number, body: RideStopDto) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.post<RideStopDto, RideResponseDto>(
      `/api/rides/${rideId}/stop`,
      body,
      config
    );
  }

  ridePanic(rideId: number, body: RidePanicDto) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.post<RidePanicDto, RideResponseDto>(
      `/api/rides/${rideId}/panic`,
      body,
      config
    );
  }

  getRide(id: number) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.get<void, OngoingRideDto>(
      `/api/rides/ongoing`,
      config
    );
  }

  getActiveDrivers() {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
      },
    };

    return this.apiClient.get<void, DriverSummaryDto[]>(
      '/api/drivers/active',
      config
    );
  }

  createRideRequest(body: RideRequestDto) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
    };

    return this.apiClient.post<RideRequestDto, RideRequestResponseDto>(
      '/api/rides/requests',
      body,
      config
    );
  }

  startRide(rideId: number, payload: StartRideDto): Observable<ApiResponse<void>> {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json'
      },
      authenticated: true,
    }
    return this.apiClient.post<StartRideDto, void>(`/api/rides/${rideId}/start`, payload, config);
  }

  reportRide(rideId: number, body: RideReportRequestDto) {
  const config: RequestConfig = {
    headers: {
      contentType: 'application/json',
    },
    authenticated: true,
  };

  return this.apiClient.post<RideReportRequestDto, void>(
    `/api/rides/${rideId}/report`,
    body,
    config
  );
}

rateRide(rideId: number, body: RatingRequestDto) {
  const config: RequestConfig = {
    headers: {
      contentType: 'application/json',
    },
    authenticated: true,
  };

  return this.apiClient.post<RatingRequestDto, void>(
    `/api/rating/${rideId}`,
    body,
    config
  );
}

finishOngoingRide() {
  const config: RequestConfig = {
    headers: {
      accept: 'application/json',
    },
    authenticated: true,
  };

  return this.apiClient.post<void, OngoingRideDto>(
    '/api/rides/finish-ongoing',
    undefined,
    config
  );
}

getOngoingRide(): Observable<ApiResponse<ActiveRideDto>> {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json'
      },
      authenticated: true,
    };
    return this.apiClient.get<void, ActiveRideDto>(`/api/rides/ongoing`, config);
  }

getDriverRideHistory(dateFrom?: Date | null, dateTo?: Date | null, sort: 'asc' | 'desc' = 'asc') {
    const queryParams: any = {
      sort: sort.toUpperCase()
    };

    if (dateFrom) {
      queryParams.dateFrom = dateFrom.toISOString();
    }
    if (dateTo) {
      queryParams.dateTo = dateTo.toISOString();
    }

    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json'
      },
      params: queryParams,
      authenticated: true,
    };

    return this.apiClient.get<void, RideHistoryDto[]>(
      `/api/rides/driver/history`,
      config
    );
  }

  getScheduledRides(userId: number) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.get<void, RideHistoryDto[]>(`/api/rides/user/${userId}/scheduled`, config);
  }

  cancelScheduledRide(rideId: number, dto: RideCancellationDto) {
    const config: RequestConfig = {
      headers: {
        accept: 'application/json',
        contentType: 'application/json',
      },
      authenticated: true,
    };

    return this.apiClient.post<RideCancellationDto, RideHistoryDto>(`/api/rides/scheduled/${rideId}/cancel`, dto, config);
  }
}

File: voom-app/app/src/main/java/inc/visor/voom/app/shared/api/RideApi.java
package inc.visor.voom.app.shared.api;

import java.util.List;

import inc.visor.voom.app.driver.dto.ActiveRideDto;
import inc.visor.voom.app.driver.dto.StartRideDto;
import inc.visor.voom.app.shared.dto.RideHistoryDto;
import inc.visor.voom.app.shared.dto.StartScheduledRideDto;
import inc.visor.voom.app.shared.dto.ride.RideCancellationDto;
import inc.visor.voom.app.shared.dto.ride.RideResponseDto;
import inc.visor.voom.app.shared.dto.ride.RideStopDto;
import inc.visor.voom.app.shared.dto.route.RouteEstimateRequestDto;
import inc.visor.voom.app.shared.dto.route.RouteEstimateResponseDto;
import inc.visor.voom.app.user.favorite_route.dto.FavoriteRouteDto;
import inc.visor.voom.app.user.home.dto.CreateFavoriteRouteDto;
import inc.visor.voom.app.user.home.dto.RideRequestDto;
import inc.visor.voom.app.user.home.dto.RideRequestResponseDto;
import inc.visor.voom.app.user.tracking.dto.ComplaintRequestDto;
import inc.visor.voom.app.user.tracking.dto.RatingRequestDto;
import inc.visor.voom.app.user.tracking.dto.RidePanicDto;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Query;
import retrofit2.http.Path;

public interface RideApi {

    @POST("/api/rides/requests")
    Call<RideRequestResponseDto> createRideRequest(
            @Body RideRequestDto payload
    );

    @GET("/api/rides/ongoing")
    Call<ActiveRideDto> getOngoingRide();

    @GET("/api/rides/driver/history")
    Call<List<RideHistoryDto>> getDriverRideHistory(@Query("dateFrom") String dateFrom,
                                                    @Query("dateTo") String dateTo,
                                                    @Query("sort") String sort);
    @POST("/api/rides/favorites")
    Call<Void> createFavoriteRoute(@Body CreateFavoriteRouteDto dto);

    @GET("/api/rides/favorites")
    Call<List<FavoriteRouteDto>> getFavoriteRoutes();

    @DELETE("/api/rides/favorites/{id}")
    Call<Void> deleteFavoriteRoute(@Path("id") long routeId);

    @POST("/api/rides/scheduled/{rideId}")
    Call<Void> startScheduleRide(
            @Path("rideId") long rideId,
            @Body StartScheduledRideDto payload
    );

    @POST("/api/routes")
    Call<RouteEstimateResponseDto> getRouteEstimate(@Body RouteEstimateRequestDto payload);


    @POST("/api/rating/{rideId}")
    Call<Void> rateRide(@Path("rideId") Long rideId, @Body RatingRequestDto body);

    @POST("/api/complaints/ride/{rideId}")
    Call<Void> reportRide(@Path("rideId") Long rideId, @Body ComplaintRequestDto body);

    @POST("/api/rides/{rideId}/panic")
    Call<Void> panic(@Path("rideId") Long rideId, @Body RidePanicDto body);

    @POST("/api/rides/{rideId}/cancel")
    Call<Void> cancel(@Path("rideId") Long rideId, @Body RideCancellationDto body);
    @POST("/api/rides/finish-ongoing")
    Call<Void> finishOngoing();

    @GET("/api/rides/ongoing/driver/{driverId}")
    Call<ActiveRideDto> getOngoingByDriverId(@Path("driverId") long driverId);

    @POST("/api/rides/{rideId}/start")
    Call<Void> startRide(
            @Path("rideId") long rideId,
            @Body StartRideDto dto
    );

    @POST("/api/rides/{id}/stop")
    Call<RideResponseDto> stopRide(@Path("id") Long id, @Body RideStopDto dto);
}

